File: 2 - Stack/C/stack.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#define STR_MAX 100

typedef struct {
    char** data;   // Pointer to an array of char pointers (strings)
    int top;
    int capacity;
} Stack;

// Initialize the dynamic stack
void initDynStack(Stack* s, int capacity) {
    s->data = (char**)malloc(capacity * sizeof(char*));
    s->top = -1;
    s->capacity = capacity;
}

// Check if the stack is full
int isFull(const Stack* s) {
    return s->top == s->capacity - 1;
}

// Check if the stack is empty
int isEmpty(const Stack* s) {
    return s->top == -1;
}

// Resize the stack
void resize(Stack* s) {
    s->capacity *= 2;
    s->data = (char**)realloc(s->data, s->capacity * sizeof(char*));
}

// Push a string onto the stack
void push(Stack* s, const char* value) {
    if (isFull(s)) {
        resize(s);
    }
    s->top++;
    s->data[s->top] = (char*)malloc(STR_MAX * sizeof(char));  // Allocate memory for the string
    strcpy(s->data[s->top], value);
    printf("Pushed %s onto stack.\n", value);
}

// Pop a string from the stack
void pop(Stack* s) {
    if (isEmpty(s)) {
        printf("Stack is EMPTY! Cannot pop.\n");
        return;
    }
    char* poppedValue = s->data[s->top];
    s->top--;
    printf("Popped %s from stack.\n", poppedValue);
    free(poppedValue);  // Free the memory for the popped string
}

// Get the top value without removing it
const char* topValue(const Stack* s) {
    if (isEmpty(s)) {
        printf("Stack is EMPTY! No top value.\n");
        return NULL;
    }
    return s->data[s->top];
}

// Peek a string at the n-th position from the top
const char* peek(const Stack* s, int n) {
    if (isEmpty(s)) {
        printf("Stack is EMPTY! No element to peek.\n");
        return NULL;
    }
    int index = s->top - (n - 1);
    if (index < 0 || index > s->top) {
        printf("Invalid peek(%d). Not enough elements.\n", n);
        return NULL;
    }
    return s->data[index];
}

// Display the contents of the stack
void display(const Stack* s) {
    if (isEmpty(s)) {
        printf("Stack is EMPTY!\n");
        return;
    }
    printf("Stack contents (top -> bottom):\n");
    for (int i = s->top; i >= 0; i--) {
        printf("[%d]: %s\n", i + 1, s->data[i]);
    }
    printf("\n");
}

// Destroy the stack and free all memory
void destroyDynStack(Stack* s) {
    for (int i = 0; i <= s->top; i++) {
        free(s->data[i]);
    }
    free(s->data);  // Free the array of pointers
    s->data = NULL;
    s->top = -1;
    s->capacity = 0;
    printf("Stack destroyed.\n");
}

int main() {
    Stack books;
    initDynStack(&books, 5);

    push(&books, "Mathematics");
    push(&books, "Calculus");
    push(&books, "Physics");
    push(&books, "Web Design");
    push(&books, "Data Structure");

    display(&books);

    printf("Top of stack: %s\n", topValue(&books));
    printf("Peek(3): %s\n", peek(&books, 2));

    pop(&books);

    destroyDynStack(&books);

    return 0;
}

File: 2 - Stack/Python/stack.py
class Stack:
    def __init__(self, capacity):
        """
        Initialize the dynamic stack with a given capacity.
        
        :param capacity: Initial capacity of the stack
        """
        self.data = []
        self.top = -1
        self.capacity = capacity

    def is_full(self):
        """
        Check if the stack is full.
        
        :return: True if the stack is full, False otherwise
        """
        return self.top == self.capacity - 1

    def is_empty(self):
        """
        Check if the stack is empty.
        
        :return: True if the stack is empty, False otherwise
        """
        return self.top == -1

    def resize(self):
        """
        Resize the stack by doubling its capacity.
        """
        self.capacity *= 2

    def push(self, value):
        """
        Push a string onto the stack.
        
        :param value: The string to be pushed onto the stack
        """
        if self.is_full():
            self.resize()
        self.top += 1
        if self.top < len(self.data):
            self.data[self.top] = value
        else:
            self.data.append(value)
        print(f"Pushed {value} onto stack.")

    def pop(self):
        """
        Pop a string from the stack.
        
        :return: The popped string, or None if the stack is empty
        """
        if self.is_empty():
            print("Stack is EMPTY! Cannot pop.")
            return None
        popped_value = self.data[self.top]
        self.top -= 1
        print(f"Popped {popped_value} from stack.")
        return popped_value

    def top_value(self):
        """
        Get the top value without removing it.
        
        :return: The top value, or None if the stack is empty
        """
        if self.is_empty():
            print("Stack is EMPTY! No top value.")
            return None
        return self.data[self.top]

    def peek(self, n):
        """
        Peek a string at the n-th position from the top.
        
        :param n: The position from the top to peek at
        :return: The string at the n-th position, or None if invalid
        """
        if self.is_empty():
            print("Stack is EMPTY! No element to peek.")
            return None
        index = self.top - (n - 1)
        if index < 0 or index > self.top:
            print(f"Invalid peek({n}). Not enough elements.")
            return None
        return self.data[index]

    def display(self):
        """
        Display the contents of the stack.
        """
        if self.is_empty():
            print("Stack is EMPTY!")
            return
        print("Stack contents (top -> bottom):")
        for i in range(self.top, -1, -1):
            print(f"[{i + 1}]: {self.data[i]}")
        print()

    def destroy(self):
        """
        Destroy the stack and free all memory.
        """
        self.data = []
        self.top = -1
        self.capacity = 0
        print("Stack destroyed.")


if __name__ == "__main__":
    books = Stack(5)

    books.push("Mathematics")
    books.push("Calculus")
    books.push("Physics")
    books.push("Web Design")
    books.push("Data Structure")

    books.display()

    print(f"Top of stack: {books.top_value()}")
    print(f"Peek(3): {books.peek(2)}")

    books.pop()

    books.destroy()